tests/Unit/Evolution/Systems/ScalarWave/Test_Constraints.cpp

```diff
@@ -0,0 +1,201 @@
+      static_cast<tnsr::ij<DataVector, Dim, Frame::Inertial> (*)(
+          const tnsr::ij<DataVector, Dim, Frame::Inertial>&)>(
+          &ScalarWave::two_index_constraint<Dim>),
+      "Constraints", "two_index_constraint", {{{-10.0, 10.0}}}, used_for_size,
+      1.0e-12);
+}
+
+// Test the return-by-value four-index constraint function using random values
+template <size_t Dim>
+void test_one_index_constraint_random(
+    const DataVector& used_for_size) noexcept {
+  pypp::check_with_random_values<1>(
+      static_cast<tnsr::i<DataVector, Dim, Frame::Inertial> (*)(
+          const tnsr::i<DataVector, Dim, Frame::Inertial>&,
+          const tnsr::i<DataVector, Dim, Frame::Inertial>&)>(
+          &ScalarWave::one_index_constraint<Dim>),
+      "Constraints", "one_index_constraint", {{{-10.0, 10.0}}}, used_for_size);
+}
+
+template <typename Solution>
+void test_constraints_and_compute_tags_analytic(
+    const Solution& solution, const size_t grid_size_each_dimension,
+    const std::array<double, 3>& lower_bound,
+    const std::array<double, 3>& upper_bound,
+    const double error_tolerance) noexcept {
+  // Shorter names for tags.
+  constexpr size_t dim = 3;
+
+  // Check vs. time-independent analytic solution
+  // Set up grid
+  Mesh<3> mesh{grid_size_each_dimension, Spectral::Basis::Legendre,
+               Spectral::Quadrature::GaussLobatto};
+  const size_t data_size = mesh.number_of_grid_points();
+
+  using Affine = domain::CoordinateMaps::Affine;
+  using Affine3D =
+      domain::CoordinateMaps::ProductOf3Maps<Affine, Affine, Affine>;
+  const auto coord_map =
+      domain::make_coordinate_map<Frame::Logical, Frame::Inertial>(Affine3D{
+          Affine{-1.0, 1.0, lower_bound[0], upper_bound[0]},
+          Affine{-1.0, 1.0, lower_bound[1], upper_bound[1]},
+          Affine{-1.0, 1.0, lower_bound[2], upper_bound[2]},
+      });
+
+  // Set up coordinates
+  const auto x_logical = logical_coordinates(mesh);
+  const auto x = coord_map(x_logical);
+  // Arbitrary time for time-independent solution.
+  const double t = 0.;
+
+  // Evaluate analytic solution
+  const auto vars = solution.variables(
+      x, t,
+      tmpl::list<ScalarWave::Pi, ScalarWave::Phi<dim>, ScalarWave::Psi>{});
+  const auto& psi = get<ScalarWave::Psi>(vars);
+  const auto& phi = get<ScalarWave::Phi<dim>>(vars);
+  const auto& pi = get<ScalarWave::Pi>(vars);
+  // Compute derivatives d_phi, d_pi, and d_gauge_function numerically
+  // First, prepare
+  using variables_tags_to_differentiate =
+      tmpl::list<ScalarWave::Psi, ScalarWave::Pi, ScalarWave::Phi<3>>;
+  Variables<variables_tags_to_differentiate> local_vars(data_size);
+  get<ScalarWave::Psi>(local_vars) = psi;
+  get<ScalarWave::Pi>(local_vars) = pi;
+  get<ScalarWave::Phi<dim>>(local_vars) = phi;
+  // Second, compute derivatives
+  const auto local_derivs =
+      partial_derivatives<variables_tags_to_differentiate,
+                          variables_tags_to_differentiate>(
```

Comment:
you don't need the second template parameter, it will be inferred

Reply:
removed